### PR TITLE
SDIT-1798 Offender assessments causing database lock and hikari timeouts

### PR DESCRIPTION
**Pull request by steverendell**

### CODE REVIEW

File: src/main/java/uk/gov/justice/hmpps/prison/api/resource/OffenderAssessmentResource.java [hunk 1]
```diff
@@ -92,13 +91,11 @@ public List<Assessment> getOffenderAssessmentsAssessmentCode(
...
-        @RequestBody @Parameter(description = "The required offender numbers (mandatory)", required = true) final List<String> offenderList,
+        @RequestBody @NotEmpty(message = "List of Offender Ids must be provided.") @Parameter(description = "The required offender numbers (mandatory)", required = true) final List<String> offenderList,
         @RequestParam(value = "latestOnly", required = false, defaultValue = "true") @Parameter(description = "Returns only assessments for the current sentence if true, otherwise assessments for all previous sentences are included") final Boolean latestOnly,
         @RequestParam(value = "activeOnly", required = false, defaultValue = "true") @Parameter(description = "Returns only active assessments if true, otherwise inactive and pending assessments are included") final Boolean activeOnly,
         @RequestParam(value = "mostRecentOnly", required = false) @Parameter(description = "Returns only the last assessment per sentence if true, otherwise all assessments for the booking are included") final Boolean mostRecentOnly
     ) {
-        validateOffenderList(offenderList);
```
Comment (steverendell) [inline src/main/java/uk/gov/justice/hmpps/prison/api/resource/OffenderAssessmentResource.java:100]: using annotation instead

File: src/main/java/uk/gov/justice/hmpps/prison/api/resource/OffenderAssessmentResource.java [hunk 2]
```diff
@@ -152,11 +151,14 @@ public AssessmentDetail getOffenderCsraAssessment(
...
+    public List<Assessment> getAssessments(
+        @RequestParam("offenderNo") @Parameter(description = "The required offender numbers Ids (mandatory)", required = true) @NotEmpty(message = "List of Offender Ids must be provided.") final List<String> offenderList,
+        @RequestParam(value = "latestOnly", required = false, defaultValue = "true") @Parameter(description = "Returns only assessments for the current sentence if true, otherwise assessments for all previous sentences are included") final Boolean latestOnly,
+        @RequestParam(value = "activeOnly", required = false, defaultValue = "true") @Parameter(description = "Returns only active assessments if true, otherwise inactive and pending assessments are included") final Boolean activeOnly,
+        @RequestParam(value = "mostRecentOnly", required = false) @Parameter(description = "Returns only the last assessment per sentence if true, otherwise all assessments for the booking are included") final Boolean mostRecentOnly) {
```
Comment (steverendell) [inline src/main/java/uk/gov/justice/hmpps/prison/api/resource/OffenderAssessmentResource.java:158]: reformatted to 1 parameter per line

File: src/main/java/uk/gov/justice/hmpps/prison/api/resource/OffenderAssessmentResource.java [hunk 3]
```diff
@@ -218,8 +223,11 @@ public ResponseEntity<Map<String, Long>> createCategorisation(@Valid @RequestBod
     @PreAuthorize("hasAnyRole('MAINTAIN_ASSESSMENTS','CREATE_CATEGORISATION','CREATE_RECATEGORISATION')")
     @PutMapping("/category/categorise")
     @ProxyUser
-    public ResponseEntity<Void> updateCategorisation(@Valid @RequestBody @Parameter(description = "Categorisation details", required = true) final CategorisationUpdateDetail detail) {
-        inmateService.updateCategorisation(detail.getBookingId(), detail);
+    public ResponseEntity<Void> updateCategorisation(
+        @RequestParam(value = "lockTimeout", required = false, defaultValue = "false") @Parameter(description = "Whether to timeout if locked", example = "true") final Boolean lockTimeout,
+        @Valid @RequestBody @Parameter(description = "Categorisation details", required = true) final CategorisationUpdateDetail detail
+    ) {
+        inmateService.updateCategorisation(detail.getBookingId(), detail, lockTimeout);
```
Comment (steverendell) [inline src/main/java/uk/gov/justice/hmpps/prison/api/resource/OffenderAssessmentResource.java:230]: Added lockTimeout parameter to the 5 relevant categorisation endpoints

File: src/main/java/uk/gov/justice/hmpps/prison/repository/BookingRepository.java [hunk 4]
```diff
@@ -318,10 +318,8 @@ public List<ScheduledEvent> getBookingActivities(final Collection<Long> bookingI
         return activities.stream().map(ScheduledEventDto::toScheduledEvent).collect(Collectors.toList());
     }
 
-    private final static int lockWaitTime = 10;
```
Comment (steverendell) [inline src/main/java/uk/gov/justice/hmpps/prison/repository/BookingRepository.java:321]: Moved this to an application parameter

File: src/main/java/uk/gov/justice/hmpps/prison/repository/ConditionalSqlService.kt [hunk 5]
```diff
@@ -1,20 +1,24 @@
...
+  override fun getWaitClause() = " WAIT $lockWaitTimeSeconds"
+  override fun getLockWaitTime() = lockWaitTimeSeconds
 }
 
 @Service
 @Profile("!nomis")
 class HsqlConditionalSqlService : ConditionalSqlService {
-  override fun getWaitClause(lockWaitTime: Int) = ""
+  override fun getWaitClause() = ""
+  override fun getLockWaitTime() = 0
```
Comment (steverendell) [inline src/main/java/uk/gov/justice/hmpps/prison/repository/ConditionalSqlService.kt:23]: need conditional bean as hsql does not support the "for update wait n" select syntax

File: src/main/java/uk/gov/justice/hmpps/prison/repository/OffenderCurfewRepository.java [hunk 6]
```diff
@@ -221,37 +219,34 @@ public void resetCurfew(long curfewId) {
 
     private void getCurfewLock(long curfewId) {
         getLock(OffenderCurfewRepositorySql.GET_OFFENDER_CURFEW_LOCK,
-            createParams("curfewId", curfewId, "waitSeconds", lockWaitTime),
+            createParams("curfewId", curfewId),
```
Comment (steverendell) [inline src/main/java/uk/gov/justice/hmpps/prison/repository/OffenderCurfewRepository.java:222]: removed redundant param

File: src/main/java/uk/gov/justice/hmpps/prison/repository/sql/InmateRepositorySql.kt [hunk 7]
```diff
@@ -736,7 +736,7 @@ enum class InmateRepositorySql(val sql: String) {
     """,
   ),
 
-  UPDATE_CATEORY_NEXT_REVIEW_DATE(
+  UPDATE_CATEGORY_NEXT_REVIEW_DATE(
```
Comment (steverendell) [inline src/main/java/uk/gov/justice/hmpps/prison/repository/sql/InmateRepositorySql.kt:739]: typo

File: src/main/java/uk/gov/justice/hmpps/prison/service/InmateService.java [hunk 8]
```diff
@@ -464,15 +464,11 @@ private boolean isCalculatedCsra(final AssessmentDto assessmentDto) {
...
-    private List<OffenderCategorise> doGetOffenderCategorisations(final String agencyId, final Set<Long> bookingIds, final boolean latestOnly) {
         final List<OffenderCategorise> results = new ArrayList<>();
         if (!CollectionUtils.isEmpty(bookingIds)) {
             final var batch = Lists.partition(new ArrayList<>(bookingIds), maxBatchSize);
             batch.forEach(offenderBatch -> {
-                final var categorisations = repository.getOffenderCategorisations(offenderBatch, agencyId, latestOnly);
+                final var categorisations = repository.getOffenderCategorisations(offenderBatch, null, latestOnly);
                 results.addAll(categorisations);
             });
         }
```
Comment (steverendell) [inline src/main/java/uk/gov/justice/hmpps/prison/service/InmateService.java:474]: Inlined this due to compiler warning

File: src/main/java/uk/gov/justice/hmpps/prison/service/InmateService.java [hunk 9]
```diff
@@ -516,7 +512,7 @@ public List<OffenderCategorise> getCategory(final String agencyId, final Categor
         };
     }
 
-    @VerifyBookingAccess(overrideRoles = {"MAINTAIN_ASSESSMENTS"}, accessDeniedError = true)
+    @VerifyBookingAccess(overrideRoles = {"MAINTAIN_ASSESSMENTS"})
```
Comment (steverendell) [inline src/main/java/uk/gov/justice/hmpps/prison/service/InmateService.java:515]: accessDeniedError setting now redundant

**Review by andymarke — APPROVED** [review]
LGTM

**Review by andymarke — APPROVED** [review]
LGTM

**Pull request merged**